### PR TITLE
net: conn_mgr: generic wifi_mgmt connectivity backend

### PR DESCRIPTION
**Pull request by JordanYates**
As connectivity backends need to be bound in the same file that the
`net_if` is created in, define a common backend type for WiFi modems.
All WiFi modems should be controllable through the `wifi_mgmt` API, so
there should be no need for dedicated context.

When enabled, the particular implementation to be used is chosen through
`CONNECTIVITY_WIFI_MGMT_IMPL`. For now, the only choice is an
application defined backend.

### CODE REVIEW

File: include/zephyr/net/conn_mgr/connectivity_wifi_mgmt.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+/* Includes ------------------------------------------------------------------*/
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * Context type for generic WIFI_MGMT connectivity backend.
+ */
+#define CONNECTIVITY_WIFI_MGMT_CTX_TYPE void *
```
[the remark this answers is not in the document]
Comment (JordanYates) [reply include/zephyr/net/conn_mgr/connectivity_wifi_mgmt.h:24]: Wrapping the type in brackets results in build errors due to using the define as `static CONN_MGR_CONN_IMPL_GET_CTX_TYPE(conn_id)`  in `CONN_MGR_BIND_CONN_INST`.

File: drivers/wifi/esp32/src/esp_wifi_drv.c [hunk 2]
```diff
@@ -674,3 +676,8 @@ NET_DEVICE_DT_INST_DEFINE(0,
 		&esp32_data, NULL, CONFIG_WIFI_INIT_PRIORITY,
 		&esp32_api, ETHERNET_L2,
 		NET_L2_GET_CTX_TYPE(ETHERNET_L2), NET_ETH_MTU);
+
+#if CONFIG_NET_CONNECTION_MANAGER_CONNECTIVITY_WIFI_MGMT
```
[the remark this answers is not in the document]
Comment (JordanYates) [reply drivers/wifi/esp32/src/esp_wifi_drv.c]: done

File: include/zephyr/net/conn_mgr/connectivity_wifi_mgmt.h [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+/**
+ * @file
+ * @brief Connectivity implementation for drivers exposing the wifi_mgmt API
+ */
+
+#ifndef ZEPHYR_INCLUDE_CONN_MGR_CONNECTIVITY_WIFI_MGMT_H_
+#define ZEPHYR_INCLUDE_CONN_MGR_CONNECTIVITY_WIFI_MGMT_H_
+
+/* Includes ------------------------------------------------------------------*/
```
[the remark this answers is not in the document]
Comment (JordanYates) [reply include/zephyr/net/conn_mgr/connectivity_wifi_mgmt.h]: done

File: include/zephyr/net/conn_mgr/connectivity_wifi_mgmt.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (jukkar) [reply include/zephyr/net/conn_mgr/connectivity_wifi_mgmt.h:22]: Ok, never mind then.

File: drivers/wifi/esp32/src/esp_wifi_drv.c [hunk 4]
```diff
@@ -672,3 +674,8 @@ NET_DEVICE_DT_INST_DEFINE(0,
 		&esp32_data, NULL, CONFIG_WIFI_INIT_PRIORITY,
 		&esp32_api, ETHERNET_L2,
 		NET_L2_GET_CTX_TYPE(ETHERNET_L2), NET_ETH_MTU);
+
+#if defined(CONFIG_NET_CONNECTION_MANAGER_CONNECTIVITY_WIFI_MGMT)
+CONN_MGR_CONN_DECLARE_PUBLIC(CONNECTIVITY_WIFI_MGMT);
+CONN_MGR_BIND_CONN(Z_DEVICE_DT_DEV_ID(DT_DRV_INST(0)), CONNECTIVITY_WIFI_MGMT);
+#endif
```
Comment (jukkar) [inline drivers/wifi/esp32/src/esp_wifi_drv.c:681]: I wonder if we could introduce a new macro
```
#define CONN_MGR_BIND_CONN_WIFI(dev) \
	COND_CODE_1(CONFIG_NET_CONNECTION_MANAGER_CONNECTIVITY_WIFI_MGMT, \
		(CONN_MGR_CONN_DECLARE_PUBLIC(CONNECTIVITY_WIFI_MGMT); \
		CONN_MGR_BIND_CONN(dev, CONNECTIVITY_WIFI_MGMT)))
```
And then here we could just do
```
CONN_MGR_BIND_CONN_WIFI(Z_DEVICE_DT_DEV_ID(DT_DRV_INST(0));
```
This way we do not need to repeat the boilerplate code in every driver.

File: drivers/wifi/esp32/src/esp_wifi_drv.c [hunk 5]
```diff
@@ -672,3 +674,8 @@ NET_DEVICE_DT_INST_DEFINE(0,
 		&esp32_data, NULL, CONFIG_WIFI_INIT_PRIORITY,
 		&esp32_api, ETHERNET_L2,
 		NET_L2_GET_CTX_TYPE(ETHERNET_L2), NET_ETH_MTU);
+
+#if defined(CONFIG_NET_CONNECTION_MANAGER_CONNECTIVITY_WIFI_MGMT)
+CONN_MGR_CONN_DECLARE_PUBLIC(CONNECTIVITY_WIFI_MGMT);
```
Comment (rlubos) [inline drivers/wifi/esp32/src/esp_wifi_drv.c:679]: I think it'd make sense to use `CONN_MGR_CONN_DECLARE_PUBLIC()` in a public header (`connectivity_wifi_mgmt.h`) where the context type is defined (as suggested in [docs](https://docs.zephyrproject.org/latest/connectivity/networking/conn_mgr/implementation.html#declaring-an-implementation-publicly)).

**Review by rlubos — COMMENTED** [review]
Thanks, one step closer to getting the actual conn manager implementation for Wi-Fi (we should come up with something generic based on `wifi_mgmt` at some point).

File: drivers/wifi/esp32/src/esp_wifi_drv.c (hunk 4 above)
Comment (JordanYates) [reply drivers/wifi/esp32/src/esp_wifi_drv.c]: Done

File: drivers/wifi/esp32/src/esp_wifi_drv.c (hunk 5 above)
Comment (JordanYates) [reply drivers/wifi/esp32/src/esp_wifi_drv.c]: Resolved as part of @jukkar's suggestion, although not in this way.

**Comment by JordanYates** [comment]
> Thanks, one step closer to getting the actual conn manager implementation for Wi-Fi (we should come up with something generic based on wifi_mgmt at some point).

That is the next step. Unfortunately completely generic is not possible as the location of the network credentials won't be common, but we can add another choice of implementations for that.

**Review by jukkar — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Comment by glarsennordic** [comment]
Sorry I missed reviewing this -- Looks good, though, thanks for doing it

**Pull request merged**